File: client/src/api/mcp.ts
// MCP server management — wraps /api/mcp/* (Claude Code user-scope servers).
import { api } from "./client";

export interface MCPServer {
  name: string;
  type: string;        // "stdio" | "http" | "sse"
  command: string;
  args: string[];
  url: string;
  env_keys: string[];
}

export async function listMCPServers(): Promise<MCPServer[]> {
  const { data } = await api.get<{ servers: MCPServer[] }>("/mcp/servers");
  return data.servers;
}

export async function addMCPServer(name: string, config: Record<string, unknown>): Promise<void> {
  await api.post("/mcp/servers", { name, config });
}

export async function removeMCPServer(name: string): Promise<void> {
  await api.delete(`/mcp/servers/${encodeURIComponent(name)}`);
}

File: server/app/routers/mcp.py
"""MCP server management endpoints (Claude Code user-scope servers)."""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.core.security import require_user
from app.services import mcp_config

router = APIRouter()


class AddServerBody(BaseModel):
    name: str = Field(..., min_length=1)
    config: dict[str, Any]


def _exec(call):
    try:
        return call()
    except mcp_config.MCPError as e:
        raise HTTPException(400, str(e))


@router.get("/mcp/servers")
def list_servers(_u: str = Depends(require_user)) -> dict[str, Any]:
    return {"servers": _exec(mcp_config.list_servers)}


@router.post("/mcp/servers")
def add_server(body: AddServerBody, _u: str = Depends(require_user)) -> dict[str, Any]:
    return _exec(lambda: mcp_config.add_server(body.name, body.config))


@router.delete("/mcp/servers/{name}")
def remove_server(name: str, _u: str = Depends(require_user)) -> dict[str, Any]:
    return _exec(lambda: mcp_config.remove_server(name))
